### ISSUE TITLE
ETEO Algorithm and data

### ISSUE THREAD

**Issue by DeepAnonymous** [TradeMaster-NTU/TradeMaster#94]
Hi,

I am trying to figure out ETEO algorithm for OE. However, could you please provide the source paper of ETEO algorithm, an arXiv link should be helpful.

 Also, could you please help to explain the data used in ETEO algorithm, e.g. bids_distance_x, bids_notional_x, bids_cancel_notional_x, bids_limit_notional_x, bids_market_notional_x, asks_distance_x, asks_notional_x, asks_cancel_notional_x, asks_limit_notional_x, asks_market_notional_x. Normally, I only notice bid ask price and volume, so it's difficult for me to collect data on my own.

Thank you for your help.

**Comment by sunshuo01** [comment]
Hi,

Here is the link of the paper https://www.ijcai.org/Proceedings/2020/627  

For the data, the current order execution environment is a realistic but complex version that use fine-grained trade data. You can check and download the data here https://drive.google.com/drive/folders/19Tk5ifPz1y8i_pJVwZFxaSueTLjz6qo3 

We also plan to provide a simple version later which use bid/ask price and volume as you expected. Please let us know if you have any further questions.

**Comment by qinmoelei** [comment]
Hi,
Thanks for the feedback. The link for the article is https://www.ijcai.org/proceedings/2020/0627.pdf.
For the data structure, ask and bid represents the side of the orderbook. asks_distance_x represents the distance of the midpoint price and price that a seller want to sell at, and bids_distance_x  represents the distance of the midpoint price and the price that a buyer want to buy. The x represents order of distance. xxx_notional_x represents the overall volume, xxx_limit_notional_x represents the volume in the limit orderbook and xxx_market_notional_x represents the volume that could be done with market maker, xxx_cancel_notional_x represents the order has been cancelled from the last timestamp to this timestamp. You should verify that xxx_notional_x =xxx_limit_notional_x +xxx_market_notional_x -xxx_cancel_notional_x, where xxx represents asks or bids.

**Comment by DeepAnonymous** [comment]
Thank you both of you @ai-gamer and @qinmoelei for your answers.

1. Regarding the paper, I did briefly read the paper before, but the author did not provide any source code. So you guys implemented this on your own? If so, why did you choose the data that complex? Why not bid/ask price and volume as we normally see in the LOB? How could we collect such kind of complicated fine-grained data from exchanges?

2. I want to collect and train data using my own data, e.g. Binance BTCUSDT, is it possible to modify your code to adapt with bid/ask price/volume only?

3. Here is another open-source OE project that I'm interested recently, https://arxiv.org/pdf/2208.06244.pdf, with PPO and DQN implementation. Would you please have a look and give your thoughts on the differences?

4. Also, do you have any implementation of TWAP/VWAP for benchmarking purpose. It's good to have such a baseline comparison. I'd be grateful if you could provide it.

**Comment by qinmoelei** [comment]
Hi,

Regarding Question 1: yes, we implemented this on our own. The reason why we use this fine-grained data is that we build a order-maker  environment instead of order-taker environment, which we already have in the high frequency trading environment(https://github.com/TradeMaster-NTU/TradeMaster/blob/1.0.0/trademaster/environments/high_frequency_trading/environment.py), if you are interested. In the order-taker enviornment, all we need is the snapshot of LOB and some OHLC features (level 2 data mentioned in the article regarding Question 3).

Regarding Question 2: the algorithms is adptive, which means that you can use the HighFrequencyTradingEnvironment I just mentioned as the RL environment and use the ETEO algorithm to solve it. However, if you want to get stick to the article, you have to modify the reward function and the avaliable action.

Regarding Question 3: I think the difference lies in the class Broker, in our ETEO environment, the broker is fixed. The order place will last for 30 mintues before it is cancelled or executed while in this article, there seems to be more options of breakers in their simulator. The reward functions seems to be the same as ETEO in some way. I might need more time to  go through their code but I only find a github link to download data. 

Regarding Question 4: Actually, the ETEO use TWAP as a baseline itself. In the article, using the training process, if the RL surpass the TWAP policy, which means that RL use less money to buy the target amount of bitcoin, then it get the reward of 1 at the end of episode.  If you run our experiments, we will show two metrics : cashleft and TWAP, where TWAP indicates the TWAP  benchmarking.

**Comment by DeepAnonymous** [comment]
Thank you!
Regarding Question 1: The order-maker env mean that the implemented algorithm will make a limited order and it will appear in the LOB waiting to be filled, while order-taker env will buy/sell shares at the market prices? Beside, which level is your data with xxx_distance_x, xxx_notional_x, xxx_limit_notional_x +xxx_market_notional_x -xxx_cancel_notional_x?

Regarding Question 2: By saying "get stick to the article, you have to modify the reward function and the available action" => you mean that your implementation is slightly different from the original paper (using a sparse reward as presented in page 3).

Regarding Question 3: I am looking forward to your further analysis regarding Broker class.

Regarding Question 4: I will play around. However, I just want to ask that, in some other repos (https://github.com/NikhilSehgal123/coinbase-execution-algorithm/blob/main/coinbase_algo.py), they compute TWAP/VWAP and use this value to decide whether to use limit or market order (`PriceIsFavorable`)? I am wondering do we have implemented this in our action space. So, the output action does not only contain the number of shares to trade, but also the order type (limit/market).

**Comment by qinmoelei** [comment]
Regarding Question 1: a) yes, but order-taker is still a little different than the market price. We eliminate the assumption of liquidity of market where most market simulators do and the price we buy/sell at depends on our volume. The larger volume, the less favourable price you will get as an order-taker. E.g. if you want to buy as an order-taker, what you should do is to buy at the least price that a seller what to sell. And if you buy out all the volume that a ask1_volume contains, then you have to buy the rest at ask2_price, which is a less favourable because it is more expensive. The dataset we use has 15 levels for all the columns you mentioned in the Question 1.

Regarding Question 2: Our high-frequency-trading environment is not for the ETEO, but for another algorithm, ddqn and the prime goal of the high frequency trading environment is to do algorithm trading instead of order execution, but it fits your requirements of only use snapshot of orderbook and OHLC. The main difference of order excution and algorithms trading is that the action itself is single-directional and force the order to be completed at the very end while the algorithm trading is double-directional and do not specify any position at the end. More specifically, if your target is to buy one BTC in order execution, all the action you can do is buying while algorithm trading could do both selling and buying. So if you already have some BTC bought and price goes very high, in the order execution, you could not sell the BTC you have and wait for the price to drop and buy it then to gain more profit while the algorithms trading could do that, which can lead to more profit. 
If you want to use the high frequency trading environment as an order execution environment, you have to do 3 things: 1) force the agent to complete the rest order when terminal and 2) constrain the action posed so that it only trade at one side. 3) modify the reward as the ETEO article described.

Regarding Question 4: So far we do is to use LOB as order-taker to pose our TWAP policy, which means that if I have 100 timestamps left to buy one BTC, what we do is buying 0.01BTC at each timestamp and the price follow the process we discussed in Question 1.

**Comment by DeepAnonymous** [comment]
Thank you, let me check more throughfully.

Just one more question, is it possible to consider order-maker-taker at the same time? Just like human do.

**Comment by qinmoelei** [comment]
Yes, I think it can be done. Actually, that is what we do in the ETEO environment, where the action has 2 dimensions: volume (positive or negative represents buy or sell) and price. If we want to buy and our price is higher than the ask1_price (which is calculated by adding the asks_distance_1 and midpoint price), then the order is executed right away with the price ask1_price if the ask1_volume is large enough.

**Comment by DeepAnonymous** [comment]
I just run the training code for 100 epochs. However, it's still not better TWAP when validation.

Train Episode: [72/100]
+--------------+--------------+-----------------+
|  Cash Left   |     TWAP     | Cash Left Ratio |
+--------------+--------------+-----------------+
| 47933.206548 | 37461.153500 |    27.954433%   |
+--------------+--------------+-----------------+
Valid Episode: [72/100]
+--------------+--------------+-----------------+
|  Cash Left   |     TWAP     | Cash Left Ratio |
+--------------+--------------+-----------------+
| 35247.463185 | 38610.495450 |    -8.710151%   |
+--------------+--------------+-----------------+

Should I train for even longer?

**Comment by qinmoelei** [comment]
Training longer is one option, but it shows that there might be chance that the method of ETEO could not be generalized very well on other datasets than the datasts used in the articles. What you can do is tuning the parameters of trainer and agent and see if you can get better results. The parameter config file is under the folder `configs/order_execution/order_execution_BTC_eteo_eteo_adam_mse.py` and here is the link(https://github.com/TradeMaster-NTU/TradeMaster/blob/1.0.0/configs/order_execution/order_execution_BTC_eteo_eteo_adam_mse.py)

**Comment by DeepAnonymous** [comment]
I use default settings and BTC dataset that you provided. What is the best result in your side? Just want to make sure I can replicate what you've done.

**Comment by qinmoelei** [comment]
The number is basically the same, in the training it can be better than the TWAP and cash left ratio is about 30% while in the valid and test dataset it is not better than TWAP.

**Comment by DeepAnonymous** [comment]
In the paper, they said that they used a "**one-year** millisecond TAQ data of 14 stocks from WRDS and reconstruct it into the LOB". The training set is about 9-month data, while we only have about 2-week data. Do you think we are having overfitting problem due to small amount of training data.

**Comment by qinmoelei** [comment]
Yes, it is very possible to happen. Actually, the same thing happens to our high frequency trading algorithms. With the use of one-month data, we achieve state of art result and make a load of money while here in our tutorial, we cannot upload dataset that large and therefore only use 1 day data. The result is poor.

**Comment by DeepAnonymous** [comment]
The latest update with `action[1]=action[1]+self.data["midpoint"]` => `action[1]=action[1]` in the `eteo_environment.py` does not seem to work. The network did not update.
Train Episode: [1/2]
+--------------+--------------+-------------------+
| Cash Left 1  |    TWAP 1    | Cash Left Ratio 1 |
+--------------+--------------+-------------------+
| **37461.153500** | 41059.122004 |     -**8.762897**%    |
+--------------+--------------+-------------------+
Valid Episode: [1/2]
+--------------+--------------+-------------------+
| Cash Left 1  |    TWAP 1    | Cash Left Ratio 1 |
+--------------+--------------+-------------------+
| 38610.495450 | 37244.542270 |     3.667526%     |
+--------------+--------------+-------------------+
Train Episode: [2/2]
+--------------+--------------+-------------------+
| Cash Left 1  |    TWAP 1    | Cash Left Ratio 1 |
+--------------+--------------+-------------------+
| **37461.153500** | 41059.122004 |     -**8.762897**%    |
+--------------+--------------+-------------------+
Valid Episode: [2/2]
+--------------+--------------+-------------------+
| Cash Left 1  |    TWAP 1    | Cash Left Ratio 1 |
+--------------+--------------+-------------------+
| 38610.495450 | 37244.542270 |     3.667526%     |
+--------------+--------------+-------------------+

But could you please explain what is the point of adding `+self.data["midpoint"]` in the previous code.

**Comment by qinmoelei** [comment]
Sorry for the late reply and this is a very good question.
The point is the action space has 2 dimensions: price and volume. While the volume can be ranged from [0,1] to be valid, which fits the nn output. The price has to reach certain point to make the order valid, otherwise in our environment, your order will be made but after 30 min, it will be cancelled. It is like in the real world, BTC is worth around 20000 dollars. The PPO agent will explore the price. But it will take a long time before it can sample a price around 20000 dollars and place an order that can be executed in 30 min, so one way to solve it to add the midpoint price and the agents only care the relative price, which make the order placed much easier to be executed. There are other ways like using the mean of history the price. The original article ignores this part, so we did what we think is proper.

**Comment by DeepAnonymous** [comment]
Yes, I agree. What I mean is why did you guys remove that part in the current code? Adding the midpoint price is clearly better.

**Comment by qinmoelei** [comment]
The midpoint price previous version used is the midpoint price of last timestamp, so it is not stable, which might cause the agent to learn little about the global price information. We will consider more proper way to help the agents to explore, but currently, we stick to the article which does not specify the price information normalization in the action.

**Comment by DeepAnonymous** [comment]
Thank you. I have one more problem understanding the code, please help to explain this:

- It seems to me the variable `length_keeping` is not being used.
- `data_left` is determined by dividing the `df` by `3600` (`data_left = [(len(self.df.index.unique()) - self.time_frame) / 3600]`). Does this mean if my dataset has 72000 rows, then I have `data_left = 20`? => This mean that we spread our order into 20 sub-orders within the entire timeframe of 72000 rows.
- How the batch_size affect the training process? Because I see `self.batch_size = int(get_attr(kwargs, "batch_size", 64)) `in the eteo_trainer.py, but the variable `self.batch_size` is not used.

I am asking this because normally, TWAP algorithm depends on **the number of sub-orders** and **timeframe** (15 minutes, 1 hour, 2 hours, etc.). And it's hard to understand how these values are calculated in ETEO. 

FYI: I am using another dataset spanning around 10 days LOB. That's why using the entire df to decide data_left does not make sense.

**Comment by qinmoelei** [comment]
> * It seems to me the variable `length_keeping` is not being used.

The `length_keeping` in the environment is used as `self.order_length`. In this environment, I create a list to record the order that the agent made, it is called `self.order_history` and if `self.order_history`'s length is larger than the `self.order_length`, then we pop the first order in the `self.order_history`, meaning we cancel the first order, if some order in `self.order_history` is executed. Then the volume of the order in the `self.order_history` will be changed to 0.

> * `data_left` is determined by dividing the `df` by `3600` (`data_left = [(len(self.df.index.unique()) - self.time_frame) / 3600]`). Does this mean if my dataset has 72000 rows, then I have `data_left = 20`? => This mean that we spread our order into 20 sub-orders within the entire timeframe of 72000 rows.

This is just for normalization, indicating how many hours we have left. A more general way of doing this is to use `data_left = [(len(self.df.index.unique()) - self.time_frame) / len(self.df.index.unique())]` to calculate the left time as ratio. To make it clear, you make a sub-order every time stamp.


> * How the batch_size affect the training process? Because I see `self.batch_size = int(get_attr(kwargs, "batch_size", 64)) `in the eteo_trainer.py, but the variable `self.batch_size` is not used.

Sorry for the misunderstanding, `self.batch_size` for ETEO is in `trademaster/agents/order_execution/eteo.py`. It is in the agent class so you should change the config for agent instead of trainer.



> FYI: I am using another dataset spanning around 10 days LOB. That's why using the entire df to decide data_left does not make sense.

That is a very good point. The next step for our tutorial is about using customed dataset. And at that time, we will modify the code and tutorial again. Thanks for your support.

**Comment by sunshuo01** [comment]
Hi DeepAnonymous and molei,

I would like to close this issue since it is becoming too long with different questions and may not be easy for other users to follow. Could you please open another issue for further discussions and make sure each issue focuses on one particular question? Thank you.
